Migrates listbox class names to .select;

seangates opened the pull request:
## Description
- Migrates all `.listbox` classnames to `.select`
- ~Updates for markup changes in the ebayui-core `ebay-select` component.~ (see bullet 5 below)
- Normalizes common selectors and rules into a `/base/select.less`
- Handles additional modifiers for DS6 select
- Supports both background and foreground SVG (removes the DS6 `::after` pseudo-element), requiring a slight markup change

## Context
Needed for DS4 being supported in ebayui-core.

## References
Fixes #222 

## Screenshots
![image](https://user-images.githubusercontent.com/105656/42726411-588f811a-8751-11e8-8360-1001a7e19441.png)
![image](https://user-images.githubusercontent.com/105656/42726414-61d11e1e-8751-11e8-99f6-7d836fa148d2.png)
![image](https://user-images.githubusercontent.com/105656/42726419-6d4d8f20-8751-11e8-9dcb-b9d2c8004b6f.png)

ianmcburnie commented:
I take it this a partial fix for #222 and then you will be filing more PRs for the rest of the changes (docs, markup, source folders, test pages, etc)?

ianmcburnie commented on docs/_includes/ds4/listbox.html:
```diff
@@ -6,145 +6,127 @@ <h2>Listbox</h2>
     {% highlight html %}
-<span class="listbox">
-    <select class="listbox__control listbox__control--borderless">
+<span class="select select--borderless\">
```
Typo. Stray `\` in class.

seangates replied:
Will fix.

seangates commented:
@ianmcburnie Not sure what you mean. This has source, docs and markup changes.

ianmcburnie commented:
But it still says "Listbox"everywhere.

seangates commented:
Ah. Gotcha. The nomenclature, not the class names. I can fix that in this PR, sure.

ianmcburnie commented:
Also, it's not following the pattern of `base`, `ds4` and `ds6` in the `src\select` folder like I started with checkbox and radio.

And the markup should support background SVG and foreground SVG. I added a comment in the issue about this.

You'll also need to update the ds4 test page with the new classnames and nomenclature.

ianmcburnie commented:
This may have got broken along the way (or we may have missed it in a prior PR).

<img width="474" alt="screen shot 2018-07-14 at 9 52 56 am" src="https://user-images.githubusercontent.com/38065/42726592-c9e54058-874b-11e8-9b9c-1c72c397035a.png">

seangates commented:
👍 Workin' on all of it.

ianmcburnie commented:
Feel free to break it out into separate PRs if you want to keep the scope of this one to just changing `.listbox` to `.select` as per title.

Also I would prefer if we leave the old `.listbox` around (and listbox.browser.json) for awhile until we are sure we want to delete it (and remove backwards compatibility for folks still using it).

seangates commented:
Okay, I've made some updates this evening (in prep for tomorrow). Let me know your thoughts, @ianmcburnie.

A couple of things to note from your previous comments:
- The listbox CSS is still intact, but no longer used for the examples
- Background SVG with `::after` pseudo-element causes problems for both disabled state and for custom icons. We would need to modify the `ebay-select` source to make the disabled icons work properly
- Everything which can be shared is now in a `/select/base/select.less` file and shared with their respective `/select/ds{#}/select.less` file.
- Added the disabled state and corresponding example for DS6, but it isn't in the DSL-provided pattern. It was the only state I added, preferring not to add the `aria-invalid="true"` nor the `[readonly]` states since there is not a great use case for those at the moment. Please let me know if you feel they should be included in DS6.

Looking forward to comments.

'Night! 💤

ianmcburnie commented:
Does this support foreground SVG for the icon? I don't see it. It seems like you have normalized on the DS6 markup for select, which is fine for this PR, but I feel like the DS4 markup is better as it supports background and foreground SVG using a real element.

seangates commented:
Current DS4 supports a foreground SVG. But, the current `ebay-select` markup does not support it. That was my question earlier: should I also modify the `ebay-select` to support a foreground icon.

ianmcburnie commented:
Yep, ebay-select will need updating for coreui v1.0.0 (so, no hurry just yet) to use the new markup.

seangates commented:
👍 Will make this change and update the PR.

seangates commented:
@ianmcburnie Made discussed changes. Please review once more. I think this is really close to final, so let me know if you find anything.

ianmcburnie commented:
Loving this:

<img width="826" alt="screen shot 2018-07-16 at 8 48 38 pm" src="https://user-images.githubusercontent.com/38065/42795525-aff250b0-8939-11e8-88e0-b64c3eb39bf0.png">

Been wanting to add some RTL examples for a while now. Nice! Eventually we will go through all components and do this. Most of them have been setup for it.

ianmcburnie commented:
Let's try and normalize the docs and examples if we can (like I did with checkbox and radio). Or leave it for a follow up PR if you prefer.

ianmcburnie commented on dist/listbox/ds4/listbox.css:
```diff
@@ -1,14 +1,36 @@
   position: relative;
 }
-span.listbox {
+.select::after {
```
Why did you change the existing DS4 implementation of listbox to use `::after`? We should just leave the existing code as is (using span or svg element) so there are no breaking changes until we decide to officially remove it.

ianmcburnie replied:
Oh. I see the listbox.css file is now getting all the new `.select` classname rules. But we should leave the existing listbox.css file in place for backwards compatibility. Let's point the new `select` module to a `select.css` only.

seangates replied:
Yup. Somehow this got generated and then I deleted the `listbox.less` which built it. So ... it's kinda a stray until I reinstate `listbox.less`.

ianmcburnie commented on src/less/listbox/ds4/listbox-base.less:
```diff
@@ -1,145 +0,0 @@
-// listbox text should NOT scale with root font-size, so all font-size defined with px
```
I don't think we had officially decided to delete this yet. Leaving it in for backwards compatibility doesn't hurt.

seangates replied:
Okay, I'll put it back.

ianmcburnie commented on src/less/select/ds4/select.less:
```diff
@@ -0,0 +1,74 @@
+
+@import "../base/select.less";
+
+.select__icon:empty {
```
Hmm, we're using `span` element selector everywhere else rather than `:empty` for the background SVG. Reason being that we keep the specificity the same when we need to add any `svg` element rules. Not sure if `:not(:empty)` would have the same specificity but I think the element selector is more readable. Maybe we should stick with consistency for now?

seangates replied:
Can you show me an example of what you mean?

But, yes, I want to make sure it's consistent.

ianmcburnie replied:
```css
// base
.select__icon {
  ..
}
// background specific
span.select__icon {
  ...
}
// foreground specific
svg.select__icon {
  ...
}
```

Rather than

```css
// base
.select__icon {
  ..
}
// background specific
.select__icon:empty {
  ...
}
// foreground specific
svg.select__icon {
  ...
}
```

Just to be consistent with elsewhere.

ianmcburnie commented on docs/static/ds6/skin.css:
```diff
@@ -2114,57 +2114,115 @@ span.inline-notice {
-  font-size: 16px;
-  height: 40px;
-  margin: 0;
+  font-size: 1rem;
```
I'm a bit nervous about the property changes, as they got lost in the diff and I can't compare them all. I would try and minimize refactoring of properties and save them for a separate PR where they are easier to diff.  I do know we want to keep select at 16px for now to match button and textbox (which are also hardcoded).

Make sure that you update and use the existing test page. There should be no regressions (right now it's in a broken state):

<img width="897" alt="screen shot 2018-07-16 at 9 06 30 pm" src="https://user-images.githubusercontent.com/38065/42796014-35ff1150-893c-11e8-9ebc-4d6318d9ad5c.png">

Compare with:

<img width="958" alt="screen shot 2018-07-16 at 9 08 10 pm" src="https://user-images.githubusercontent.com/38065/42796046-5dddf772-893c-11e8-8153-ca46858d27b8.png">

seangates replied:
I've fixed the test pages.

Also, I'm looking at why that property changed. It was not intentional as I was aiming to keep all the properties the same. In the shuffle this one might have been updated.

seangates replied:
It got changed to this: `font-size: @font-size-16;`. I've changed it back to a hardcoded `16px`.

ianmcburnie replied:
Ok great. We should be good if it was your aim to keep them the same.

ianmcburnie commented on dist/listbox/ds6/listbox.css:
```diff
@@ -1,62 +0,0 @@
-.listbox {
```
Yep. This one can be deleted as we never really had a DS6 listbox.

seangates replied:
👍

ianmcburnie commented:
Old listbox page is broken:

![screen shot 2018-07-17 at 11 43 47 am](https://user-images.githubusercontent.com/38065/42838877-c53712d0-89b6-11e8-8958-61124edd733a.png)

We should keep that working until we decide to permanently remove the DS4 listbox.

ianmcburnie commented:
You should probably duplicate that current listbox test page and repurpose it for a ds4 select test page (we'll start creating the ds6 tests soon).

ianmcburnie commented on dist/listbox/ds4/listbox.css:
```diff
@@ -16,6 +16,7 @@ select.listbox__control {
   font-family: "Helvetica Neue", Helvetica, Arial, Roboto, sans-serif;
   font-size: inherit;
   height: 40px;
+  line-height: 1.15;
```
I removed the line-height in this PR remember:

https://github.com/eBay/skin/pull/252/files#diff-a919e64bc0f0127334b9fabddcb84118L19

Any reason you are putting it back?

seangates replied:
Unintended. Will fix.

seangates replied:
I pulled that file back from `master`, not the 6.0.0. Whoops. ;-)

seangates commented:
@ianmcburnie Okay, all PR comments should be addressed now. Let me know if you have any questions or see anything.

ianmcburnie commented:
Everything looking great.... just one thing I see in Safari:

<img width="823" alt="screen shot 2018-07-18 at 7 54 46 pm" src="https://user-images.githubusercontent.com/38065/42918998-b93ff776-8ac4-11e8-91b3-b3afb9ca9bc7.png">


<img width="818" alt="screen shot 2018-07-18 at 7 57 54 pm" src="https://user-images.githubusercontent.com/38065/42919044-f37001c0-8ac4-11e8-9005-6b3a98bbd369.png">


I don't see that in DS6:

<img width="1213" alt="screen shot 2018-07-18 at 7 57 37 pm" src="https://user-images.githubusercontent.com/38065/42919037-eb905ce8-8ac4-11e8-8aa7-e2edd4ccdb99.png">

seangates commented:
@ianmcburnie It isn’t in there currently for DS6. Should I add it?

ianmcburnie commented:
I mean I don't see that issue in DS6.

seangates commented:
@ianmcburnie Oh, in DS4 it was intentional. And DS6 doesn’t specify.

seangates commented:
Ergo, I left hem as they were from before the normalization.

ianmcburnie commented:
Merging this now to unblock us. We can fix the last remaining issue before launch.

ianmcburnie commented:
I'll create a quick ticket for it.
